### PR TITLE
added top k search util

### PR DESCRIPTION
**Pull request by jessechancy**
Follow up from random search

### CODE REVIEW

File: keras_nlp/utils/text_generation_test.py [hunk 1]
```diff
@@ -208,3 +209,134 @@ def token_probability_fn(inputs):
...
+            return tf.repeat(prob, batch_size, axis=0)
+
+        batch_size = 10
+        inputs = 3 * tf.ones([batch_size, 1], dtype=tf.int32)
+        max_length = 3
+        tf.random.set_seed(42)
+        outputs = top_k_search(
+            token_probability_fn, inputs, k=2, max_length=max_length, seed=42
+        )
+        # Random sampling result with seed 42
```
Comment (chenmoneygithub) [inline keras_nlp/utils/text_generation_test.py:271]: top-k search result

File: keras_nlp/utils/text_generation_test.py [hunk 2]
```diff
@@ -208,3 +209,134 @@ def token_probability_fn(inputs):
...
+            flatten_predictions = tf.reshape(outputs[:, 1:], [-1])
+            for pred in flatten_predictions:
+                outputs_count[pred] += 1
+        self.assertAllClose(
+            outputs_count / np.sum(outputs_count),
+            [0.0, 0.0, 0.429, 0.571],
+            rtol=0.2,
+        )
+
+    def test_assert_top_k_generation_is_correct(self):
```
Comment (chenmoneygithub) [inline keras_nlp/utils/text_generation_test.py:306]: The test is to assert only top-k tokens can appear, but the name does not suggest so. Let's rename to something like `test_only_choose_from_top_k_tokens`

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Mainly looks good!

File: keras_nlp/utils/text_generation.py [hunk 3]
```diff
@@ -242,3 +242,118 @@ def token_probability_fn(inputs):
     if input_is_1d:
         return tf.squeeze(prompt)
     return prompt
+
+
+def top_k_search(
+    token_probability_fn,
+    prompt,
+    max_length,
+    k=10,
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:251]: do you mean vocab_size should be another argument? Or k should be a required arg

File: keras_nlp/utils/text_generation.py [hunk 4]
```diff
@@ -242,3 +242,118 @@ def token_probability_fn(inputs):
...
+    """
+    Text generation utility based on top k sampling.
+
+    Top k search samples the next token from the top k tokens in the
+    probability distribution provided by `token_probability_fn` and appends it
+    to the existing sequence.
+
+    Args:
+        token_probability_fn: a callable, which takes in input_sequence
+            and output the probability distribution of the next token.
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:265]: The return type should be probabilities here for this to work, if not I would need to add a softmax over it

File: keras_nlp/utils/text_generation.py [hunk 5]
```diff
@@ -242,3 +242,118 @@ def token_probability_fn(inputs):
...
+    if input_is_1d:
+        prompt = prompt[tf.newaxis, :]
+    i = prompt.shape[1]
+    while i < max_length:
+        # If the prompt has reached our desired length, exit while loop.
+        pred = token_probability_fn(prompt)
+        # If k is greater than the vocabulary size, use the entire vocabulary.
+        k = min(k, pred.shape[-1])
+        # Filter out top k tokens.
+        sorted_pred, sorted_indices = tf.math.top_k(pred, k=k, sorted=True)
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:339]: Yep doesn't need sorted, edited

File: keras_nlp/utils/text_generation.py [hunk 6]
```diff
@@ -242,3 +242,118 @@ def token_probability_fn(inputs):
...
+    token_probability_fn,
+    prompt,
+    max_length,
+    k=10,
+    seed=None,
+    end_token_id=None,
+    pad_token_id=0,
+):
+    """
+    Text generation utility based on top k sampling.
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:257]: edited

File: keras_nlp/utils/text_generation_test.py (hunk 1 above)
Comment (jessechancy) [reply keras_nlp/utils/text_generation_test.py:271]: Edited

File: keras_nlp/utils/text_generation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/utils/text_generation.py]: Talked on chat. Let's make `k` required (no default). No need to take in vocab_size, that can continue to be inferred.

File: keras_nlp/utils/text_generation.py [hunk 7]
```diff
@@ -242,3 +268,134 @@ def token_probability_fn(inputs):
...
+            tf.keras.layers.Dense(VOCAB_SIZE, activation="softmax"),
+        ]
+    )
+
+    # Define a function that outputs the next token's probability given the
+    # input sequence.
+    def token_probability_fn(inputs):
+        return model(inputs)[:, -1, :]
+
+    prompt = tf.random.uniform(shape=[5, 5], maxval=VOCAB_SIZE, dtype=tf.int64)
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:337]: This is kind of a weird prompt to show. Who would want to generate sequences after 5 totally random tokens.

Maybe we should do something like

```
BATCH_SIZE = 8
VOCAB_SIZE = 10
FEATURE_SIZE = 16
START_ID=1
END_ID=2

...

prompt = tf.fill((BATCH_SIZE, 1), START_ID)

keras_nlp.utils.top_k_search(
        token_probability_fn,
        prompt,
        k=10,
        max_length=10,
        end_token_id=END_ID)
```

We may want to update other examples if they have the same problem.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Couple comments!

File: keras_nlp/utils/text_generation.py (hunk 7 above)
Comment (mattdangerw) [reply keras_nlp/utils/text_generation.py:337]: also to be clear the ellipsis is me being lazy, not suggesting we put that in the docstring :)

File: keras_nlp/utils/text_generation.py [hunk 8]
```diff
@@ -217,16 +233,26 @@ def token_probability_fn(inputs):
...
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
+    validate_token_probability_fn(token_fn, prompt)
 
     i = prompt.shape[1]
     while i < max_length:
         # If the prompt has reached our desired length, exit while loop.
-        pred = token_probability_fn(prompt)
+        pred = token_fn(prompt)
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:255]: makes sense, edited

File: keras_nlp/utils/text_generation.py (hunk 7 above)
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py]: edited docstring

File: keras_nlp/utils/text_generation.py [hunk 9]
```diff
@@ -99,14 +112,15 @@ def greedy_search(
...
-    prompt = tf.random.uniform(shape=[5, 5], maxval=VOCAB_SIZE, dtype=tf.int64)
+    prompt = tf.fill((BATCH_SIZE, 1), START_ID)
 
     # Print the generated sequence (token ids).
     keras_nlp.utils.greedy_search(
         token_probability_fn,
         prompt,
         max_length=10,
-        end_token_id=0,)
+        end_token_id=END_ID
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:122]: nit: trailing comma

File: keras_nlp/utils/text_generation.py [hunk 10]
```diff
@@ -199,14 +222,15 @@ def random_search(
...
+    prompt = tf.fill((BATCH_SIZE, 1), START_ID)
 
     # Print the generated sequence (token ids).
-    keras_nlp.utils.random_sampling(
+    keras_nlp.utils.random_search(
         token_probability_fn,
         prompt,
         max_length=10,
-        end_token_id=0,)
+        end_token_id=END_ID
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:232]: trailing comma

File: keras_nlp/utils/text_generation.py [hunk 11]
```diff
@@ -242,3 +269,129 @@ def token_probability_fn(inputs):
...
+
+    prompt = tf.fill((BATCH_SIZE, 1), START_ID)
+
+    # Print the generated sequence (token ids).
+    keras_nlp.utils.top_k_search(
+        token_probability_fn,
+        prompt,
+        max_length=10,
+        k=4,
+        end_token_id=END_ID
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:349]: trailing comma

File: keras_nlp/utils/text_generation.py [hunk 12]
```diff
@@ -61,10 +68,13 @@ def greedy_search(
 
     Args:
         token_probability_fn: a callable, which takes in input_sequence
-            and output the probability distribution of the next token.
+            and output the probability distribution or the logits of the next
+            token.
         prompt: a list or a Tensor, can be 1D or 2D, the initial tokens to
             append generated tokens.
         max_length: int. The max length of generated text.
+        from_logits: bool. Indicates whether `token_probability_fn` outputs
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:76]: You document this but didn't actually add it? Should we just leave it off as for greedy search it is a no-op?

**Review by mattdangerw — APPROVED** [review]
Thanks! Last few nits!

**Review by chenmoneygithub — APPROVED** [review]
